Assertion failed is thrown from GrabCut with small initial GC_PR_FGD region

My program stopped working after invoking `grabCut` function with a mask with most pixels are initialized with `GC_PR_BGD` and a few pixels (4 pixels in my example) are initialized with `GC_PR_FGD`.
My program worked well when I used larger (100 pixels in my example) `GC_PR_FGD` region.
Is this behavior a bug, or just an incorrect usage of the library?

### The code used in my experiment

```grabcuttest.py
import cv2
import numpy as np
import base64

# configuration
# whether to print HTML for viewing images on paiza.io
enablePrintHTML = False
# whether to reproduce error (use small "maybe foreground" region)
enableReproduceError = True

# output an image as HTML for viewing images on paiza.io
def printHTML(fileName):
	if not enablePrintHTML:
		return
	file = open(fileName, "rb")
	data = file.read()
	file.close()
	print('<img src="data:image/png;base64,%s"><br>'
		% base64.b64encode(data).decode())

# visualize mask data for GrabCut
def visualizeGrabCut(mask):
	prFgd = cv2.compare(mask, cv2.GC_PR_FGD, cv2.CMP_EQ)
	prBgd = cv2.compare(mask, cv2.GC_PR_BGD, cv2.CMP_EQ)
	fgd = cv2.compare(mask, cv2.GC_FGD, cv2.CMP_EQ)
	bgd = cv2.compare(mask, cv2.GC_BGD, cv2.CMP_EQ)
	pr = cv2.bitwise_or(prFgd, prBgd)
	fgd = cv2.bitwise_or(fgd, prFgd)
	bgd = cv2.bitwise_or(bgd, prBgd)
	return cv2.merge([pr, bgd, fgd])

# create an image with a circle
img = np.zeros((480, 640, 3), np.uint8)
cv2.circle(img, (400, 300), 64, (255, 255, 255), -1, cv2.LINE_AA)
# output the image
cv2.imwrite("image.png", img)
printHTML("image.png")

# create mask for GrabCut
# all pixels are "maybe background"
mask = np.ones((480, 640), np.uint8) * cv2.GC_PR_BGD
if enableReproduceError:
	# except for 4 "maybe foreground" pixels inside the circle
	mask[300, 400] = cv2.GC_PR_FGD
	mask[300, 401] = cv2.GC_PR_FGD
	mask[301, 400] = cv2.GC_PR_FGD
	mask[301, 401] = cv2.GC_PR_FGD
else:
	# except for 100 "maybe foreground" pixels inside the circle
	for y in range(295, 305):
		for x in range(395, 405):
			mask[y, x] = cv2.GC_PR_FGD

# output initial status
cv2.imwrite("init_mask.png", visualizeGrabCut(mask))
printHTML("init_mask.png")

# execute GrabCut
rect = (0, 0, 640, 480)
bgdModel = np.zeros((1, 65), np.float64)
fgdModel = np.zeros((1, 65), np.float64)
cv2.grabCut(img, mask, rect, bgdModel, fgdModel, 1, cv2.GC_INIT_WITH_MASK)

# output result
cv2.imwrite("result_mask.png", visualizeGrabCut(mask))
printHTML("result_mask.png")
```

### Tested environment 1 (local)

* Windows 7 Home Premium Service Pack 1 (x64)
* Python 3.5.1

Emitted message:

```
Traceback (most recent call last):
  File "grabcuttest.py", line 62, in <module>
    cv2.grabCut(img, mask, rect, bgdModel, fgdModel, 1, cv2.GC_INIT_WITH_MASK)
cv2.error: OpenCV(3.4.2) C:\projects\opencv-python\opencv\modules\core\src\kmeans.cpp:240: error: (-215:Assertion failed) N >= K in function 'cv::kmeans'

```

### Tested environment 2 ([paiza.IO](https://paiza.io/en))

* Python3 (Python 3.6.8)

Emitted message:

```
Traceback (most recent call last):
  File "Main.py", line 62, in <module>
    cv2.grabCut(img, mask, rect, bgdModel, fgdModel, 1, cv2.GC_INIT_WITH_MASK)
cv2.error: OpenCV(4.1.0) /io/opencv/modules/core/src/kmeans.cpp:241: error: (-215:Assertion failed) N >= K in function 'kmeans'
```

--- a/modules/core/src/kmeans.cpp
+++ b/modules/core/src/kmeans.cpp
@@ -238,7 +238,7 @@ double cv::kmeans( InputArray _data, int K,
 
     attempts = std::max(attempts, 1);
     CV_Assert( data0.dims <= 2 && type == CV_32F && K > 0 );
-    CV_Assert( N >= K );
+    CV_CheckGE(N, K, "Number of clusters should be more than number of elements");
 
     Mat data(N, dims, CV_32F, data0.ptr(), isrow ? dims * sizeof(float) : static_cast<size_t>(data0.step));
 

--- a/modules/imgproc/src/grabcut.cpp
+++ b/modules/imgproc/src/grabcut.cpp
@@ -45,6 +45,8 @@
 
 using namespace cv;
 
+namespace {
+
 /*
 This is implementation of image segmentation algorithm GrabCut described in
 "GrabCut - Interactive Foreground Extraction using Iterated Graph Cuts".
@@ -228,6 +230,8 @@ void GMM::calcInverseCovAndDeterm(int ci, const double singularFix)
     }
 }
 
+} // namespace
+
 /*
   Calculate beta - parameter of GrabCut algorithm.
   beta = 1/(2*avg(sqr(||color[i] - color[j]||)))
@@ -379,12 +383,20 @@ static void initGMMs( const Mat& img, const Mat& mask, GMM& bgdGMM, GMM& fgdGMM 
         }
     }
     CV_Assert( !bgdSamples.empty() && !fgdSamples.empty() );
-    Mat _bgdSamples( (int)bgdSamples.size(), 3, CV_32FC1, &bgdSamples[0][0] );
-    kmeans( _bgdSamples, GMM::componentsCount, bgdLabels,
-            TermCriteria( CV_TERMCRIT_ITER, kMeansItCount, 0.0), 0, kMeansType );
-    Mat _fgdSamples( (int)fgdSamples.size(), 3, CV_32FC1, &fgdSamples[0][0] );
-    kmeans( _fgdSamples, GMM::componentsCount, fgdLabels,
-            TermCriteria( CV_TERMCRIT_ITER, kMeansItCount, 0.0), 0, kMeansType );
+    {
+        Mat _bgdSamples( (int)bgdSamples.size(), 3, CV_32FC1, &bgdSamples[0][0] );
+        int num_clusters = GMM::componentsCount;
+        num_clusters = std::min(num_clusters, (int)bgdSamples.size());
+        kmeans( _bgdSamples, num_clusters, bgdLabels,
+                TermCriteria( CV_TERMCRIT_ITER, kMeansItCount, 0.0), 0, kMeansType );
+    }
+    {
+        Mat _fgdSamples( (int)fgdSamples.size(), 3, CV_32FC1, &fgdSamples[0][0] );
+        int num_clusters = GMM::componentsCount;
+        num_clusters = std::min(num_clusters, (int)fgdSamples.size());
+        kmeans( _fgdSamples, num_clusters, fgdLabels,
+                TermCriteria( CV_TERMCRIT_ITER, kMeansItCount, 0.0), 0, kMeansType );
+    }
 
     bgdGMM.initLearning();
     for( int i = 0; i < (int)bgdSamples.size(); i++ )

